nixos/openafsServer: automatic extraction of authentication keys from Kerberos 5 service keytab

nessdoor opened the pull request:
## Description of changes
Contemporary OpenAFS cells authenticate both users and servers by relying on an existing Kerberos infrastructure. For legacy and compatibility reasons, OpenAFS servers do not use Kerberos 5 service keytabs directly, but instead rely on an OpenAFS-specific keyring located at `/etc/openafs/server/KeyFileExt`. This keyring must be filled and managed imperatively by way of the `asetkey` command.

To make matters more complicated, `asetkey` expects the administrator to be aware of the internal numerical encoding used by Kerberos 5 to represent encryption key types, and requires this information to be entered for each key that is added to the keyring via `asetkey add` (see [asetkey(8)](https://docs.openafs.org/Reference/8/asetkey.html)).

This PR introduces an automatic mechanism that, given an ordinary Kerberos 5 keytab, parses it for the AFS service keys (identified by a user-configured principal) and automatically constructs a `KeyFileExt` from these, declaratively and without any further user interaction. The AFS administrator only has to assign the *runtime path* of the keytab to the `services.openafsServer.authentication.keytab` option and, if needed, indicate the Kerberos 5 principal of the AFS service by setting the value of `services.openafsServer.authentication.principal`.

The legacy behavior is retained by default or by explicitly setting `services.openafsServer.authentication.keytab = null`.

This improvement constitutes a further step towards achieving complete AFS automatic deployment, easing the configuration of machines with volatile root partitions, and implementing automated tests.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

nessdoor committed:
nixos/openafsServer: add authentication.{principal/keytab} option

Allow users to set a Kerberos 5 keytab from which to automatically extract the
authentication keys necessary for AFS operations.

nessdoor committed:
doc: add release notes for the openafsServer.authentication options

spacefrogg commented on nixos/modules/services/network-filesystems/openafs/server.nix:
```diff
@@ -62,6 +62,68 @@ let
+      '';
+    in
+    pkgs.writeShellScript "asetkey-script.sh" ''
+      echo -e "rkt ${cfgAuth.keytab}\nlist -e" | ${pkgs.krb5}/bin/ktutil | \
```
Document that krb5 `ktutil` must be used here in a comment above the script to prevent future refactoring. (One could be tempted to point to `config.security.krb5.package`.)

spacefrogg commented on nixos/modules/services/network-filesystems/openafs/server.nix:
```diff
@@ -101,6 +163,39 @@ in {
 
+      authentication = {
+        principal = mkOption {
+          default = "afs/${cfg.cellName}@${lib.toUpper cfg.cellName}";
```
The Kerberos realm should be the configured default realm `security.krb5.settings.libdefaults.default_realm` and not assume that the AFS cell runs in its own realm.

nessdoor replied:
Thank you, I also took the liberty of redefining the option definition to be more fool-proof in case of refactoring of the Kerberos module

spacefrogg commented on nixos/modules/services/network-filesystems/openafs/server.nix:
```diff
@@ -296,13 +391,16 @@ in {
           ${optionalString useBuCellServDB "cp ${buCellServDB}"}
+        '' + lib.optionalString (cfg.authentication.keytab != null) ''
+          rm -f /etc/openafs/server/KeyFileExt
+          ${asetkeyScript}
```
Create `/etc/openafs/server/KeyFileExt` with proper access rights, first (e.g. 600). I am not sure whether asetkey does that on its own or just uses the current umask.

spacefrogg commented on nixos/modules/services/network-filesystems/openafs/server.nix:
```diff
@@ -101,6 +167,44 @@ in {
+           in mkOption {
+             default = "afs/${defaultRealm}@${lib.toUpper defaultRealm}";
+             defaultText =
+               literalExpression "afs/\${config.${defaultRealmPath}}>@\${toUpper config.${defaultRealmPath}}";
```
The cell name must still come from `cellName` and not assume `defaultRealmPath`. Also, `defaultRealmPath` is already in uppercase letters if it needs to be. Upcasing it unconditionally is wrong, because there might be instances where the Kerberos realm does not consist of all-upcase letters.

nessdoor replied:
Sorry, you are very much right, I didn't think it through properly...

spacefrogg commented on nixos/modules/services/network-filesystems/openafs/server.nix:
```diff
@@ -296,13 +400,16 @@ in {
           ${optionalString useBuCellServDB "cp ${buCellServDB}"}
+        '' + lib.optionalString (cfg.authentication.keytab != null) ''
+          rm -f /etc/openafs/server/KeyFileExt
+          ${asetkeyScript}
```
I just noticed that `asetkeyScript` never calls `asetkey` when the configured principal is not in the keytable or the keytable cannot be read. Thus, `KeyFileExt` never gets created. Please, make `preStart` fail with a descriptive error message in that case.

nessdoor commented:
I have modified the AWK script to exit with an error if it could not find any key for the give principal.
I have also modified the wrapping script to be more sensible to errors (`set -eo pipefail`).
I tested the whole pipeline locally, and it seems to be working as intended.

I also took the liberty of retargeting the release notes for 25.05 instead of 24.11.

spacefrogg reviewed (approved):
Looking good!

nessdoor commented:
Resolved merge conflict on release notes and `server.nix` post-reformat, and reformatted changes to comply with the new rules.
All other modifications left unchanged.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5030

nessdoor commented:
Resolved merge conflict on release notes, all other modifications left unchanged.
